test/e2e/suite/cases/usages.go

```diff
@@ -0,0 +1,114 @@
+			VolumeSource: corev1.VolumeSource{
+				CSI: &corev1.CSIVolumeSource{
+					Driver: csi.GroupName,
+					VolumeAttributes: map[string]string{
+						"csi.cert-manager.io/issuer-name":  f.Issuer.Name,
+						"csi.cert-manager.io/issuer-kind":  f.Issuer.Kind,
+						"csi.cert-manager.io/issuer-group": f.Issuer.Group,
+						"csi.cert-manager.io/dns-names":    "a.example.com,b.example.com",
+						"csi.cert-manager.io/uri-sans":     "spiffe://my-service.sandbox.cluster.local,http://foo.bar",
+						"csi.cert-manager.io/ip-sans":      "192.168.0.1,123.4.5.6",
+						"csi.cert-manager.io/duration":     "123h",
+						"csi.cert-manager.io/is-ca":        "true",
+						"csi.cert-manager.io/common-name":  "foo-bar",
+						"csi.cert-manager.io/key-usages":   "signing,digital signature,content commitment,key encipherment,key agreement,data encipherment",
+					},
+				},
+			},
+		}
+
+		testPod := &corev1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName: f.BaseName + "-",
+				Namespace:    f.Namespace.Name,
+			},
+			Spec: corev1.PodSpec{
+				Containers: []corev1.Container{
+					corev1.Container{
+						Name:    "test-container-1",
+						Image:   "busybox",
+						Command: []string{"sleep", "10000"},
+						VolumeMounts: []corev1.VolumeMount{
+							{
+								MountPath: "/tls",
+								Name:      "tls",
+							},
+						},
+					},
+				},
+				Volumes: []corev1.Volume{
+					testVolume,
+				},
+			},
+		}
+
+		By("Creating a Pod")
+		testPod, err := f.KubeClientSet.CoreV1().Pods(f.Namespace.Name).Create(testPod)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Waiting for Pod to become ready")
+		err = f.Helper().WaitForPodReady(f.Namespace.Name, testPod.Name, time.Second*10)
+		Expect(err).NotTo(HaveOccurred())
+
+		testPod, err = f.KubeClientSet.CoreV1().Pods(f.Namespace.Name).Get(testPod.Name, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Ensure the corresponding CertificateRequest should exist with the correct spec")
+		crName := util.BuildVolumeID(string(testPod.GetUID()), "tls")
+		cr, err := f.Helper().WaitForCertificateRequestReady(f.Namespace.Name, crName, time.Second)
+		Expect(err).NotTo(HaveOccurred())
+
+		err = util.CertificateRequestMatchesSpec(cr, testVolume.CSI.VolumeAttributes)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Ensure the certificate key pair exists in the pod and matches that in the CertificateRequest")
+		certData, keyData, err := f.Helper().CertificateKeyInPodPath(f.Namespace.Name, testPod.Name, "test-container-1", "/tls",
+			testVolume.CSI.VolumeAttributes)
+		Expect(err).NotTo(HaveOccurred())
+
+		err = f.Helper().CertificateKeyMatch(cr, certData, keyData)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Ensure the certificate key pair and metadata file exists in the local data directory and matches that in the CertificateRequest")
+		err = f.Helper().MetaDataCertificateKeyExistInHostPath(cr, testPod, testVolume.CSI.VolumeAttributes, testVolume.Name, "/tmp/cert-manager-csi")
```

Comment:
Does this method check usages? It isn't that clear from the test at least

Reply:
This is done in 

```
func CertificateRequestMatchesSpec(cr *cmapi.CertificateRequest, attr map[string]string) error {
```

Reply:
`CertificateRequestMatchesSpec` checks that the volume attributes have been correctly passed to the cr.
It doesn't examine the  usages in the certificate.
But that wouldn't be a good test because the usages granted don't necessarily match the usages requested.

Reply:
Indeed, and if something was going wrong here then it would be a bug in the cert-manager side, rather than the driver. I left it out because of these reasons.